src/CryptoNoteCore/Miner.cpp

```diff
@@ -95,9 +95,9 @@ namespace CryptoNote
     if (!is_mining()) {
       return true;
     }
-    freeDataset();
+    FREE_DATASET;
     bool r = request_block_template();
-    initDataset();
+    INIT_DATASET;
```

Comment:
I am afraid here this will crash each time this function is called, and it is seems to be called frequently, it won't even finish generating first dataset when new request will come.